accepted/single-file/direct.md

```diff
@@ -0,0 +1,85 @@
+
+### AppHost
+
+The AppHost bundle-handling code implements the following service, which will be registered for use by other components.
+
+```c++
+// If the requested file is available in the bundle, the function provides 
+//   - A pointer to a buffer with the contents of the file in memory 
+//   - The size of the file.
+// The function returns true if the file is found, false otherwise.
+static bool read_bundled_file(const char* name, 
+                              const void** buffer, 
+                              size_t* size);
+```
+
+### HostFxr
+
+The `Apphost` provides the a bundle-reader for `HostFxr` via the updated API:
+
+```C++
+SHARED_API int hostfxr_main_startupinfo(const int argc, 
+                                        const pal::char_t* argv[], 
+                                        const pal::char_t* host_path, 
+                                        const pal::char_t* dotnet_root,
+                                        const pal::char_t* app_path,
+                                        bool (*bundle_reader)(const char* name, 
+                                                              const void** buffer, 
+                                                              size_t* size));
+```
+
+The `HostFxr` uses this reader to read `app.runtimeconfig.json` and any other files it may need from the bundle. If the `bundle_reader` is invalid or fails to find the requested file, it falls back to look for an actual file on disk.
+
+### HostPolicy
+
+The `HostFxr` registers the `bundle_reader` with `HostPolicy` via the updated API:
+
+```C++
+struct corehost_initialize_request_t
+{
+    size_t version;
+    strarr_t config_keys;
+    strarr_t config_values;
+    bool (*bundle_reader)(const char* name, const void** buffer, size_t* size);
+};
+
+SHARED_API int __cdecl corehost_initialize(
+    const corehost_initialize_request_t *init_request, 
+    int32_t options,
+    /*out*/ corehost_context_contract *context_contract);
+```
+
+Host policy uses this reader to attempt to process the `app.deps.json` file and check for the existence of other component files within the bundle.
+
+### Runtime Components
+
+The `HostPolicy` registers the `bundle-reader` with the runtime via a new API: 
+
+```C++
+// Register a bundle-reader with CoreCLR
+//
+// Parameters:
+//  reader - The bundle-reader to use for this application
+// Returns:
+//  S_OK
+
+CORECLR_HOSTING_API(coreclr_register_bundle_reader,
+                    bool (*bundle_reader)(const char* name, 
+                                          const void** buffer, 
+                                          size_t* size));
+```
+
+`HostPolicy` communicates the location of files residing within the bundle with a special bundle-marker (`:`) in the CoreCLR properties such as `TRUSTED_PLATFORM_ASSEMBLIES`.
```

Comment:
`:` is path separator on Unix.

Reply:
The `:` can additionally be used for the bundle marker, right? We need to parse the input correctly. 
For example: `:il1.dll::il2.dll:/path/r2r.dll`. I've fixed the example in this section to use the correct separator.

Reply:
I guess you can design a scheme where you overload meaning of `:`. 

Note that the TRUSTED_PLATFORM_ASSEMBLIES list is parsed using a simple `string.Split` call in number of places today, e.g.: https://github.com/adrianoc/cecilifier/blob/71a0fe93c3da12f4eb0c969db3ff01935eca6b6b/Cecilifier.Core.Tests/Framework/Utils.cs#L12 . They are not going to work anymore if you overload meaning of `:`.

The current TPA list design has corner case bugs like https://github.com/dotnet/core-setup/issues/4141. This will add more corner case bugs to it. I am wondering whether we should fix these corner cases as well while we are on it.

Reply:
The only reason I picked overloaded meaning for `:` is to not preclude another character from use within paths. I'm happy to use a different character (say `#`) -- I'm not sure if you're suggesting the change ?

Reply:
Since we are going change the format or meaning of `TRUSTED_PLATFORM_ASSEMBLIES`, I am suggesting that it would be useful to fix the other known problems with it - where it does not allow certain characters.

It looks like you are going to have to add callbacks to read the file. Would it make sense to handle everything using the callbacks? If the callback is registered, the runtime would not look at the `TRUSTED_PLATFORM_ASSEMBLIES` list at all. It would get us onto a path to solve all escaping problems with `TRUSTED_PLATFORM_ASSEMBLIES`.

The other alternatives I can think are escaping scheme for `TRUSTED_PLATFORM_ASSEMBLIES` that allows *any* characters; or start passing around more elaborate structures than just plain strings.

Reply:
How about we say that the host passes two properties to the runtime (both strings using standard separator for the platform):
* `TRUSTED_BUNDLED_ASSEMBLIES` containing assemblies in the bundle and
* `TRUSTED_PLATFORM_ASSEMBLIES` containing remaining files on disk.

The TPA-binder will look for an assembly in `TRUSTED_BUNDLED_ASSEMBLIES` first and use a special call-back if found. Otherwise, it processes `TRUSTED_PLATFORM_ASSEMBLIES` as usual.
Similarly for `PLATFORM_RESOURCE_ROOTS`.

I think this change will cause minimal churn in the host-runtime handshake, while implementing the load-bundled-assemblies feature.

I agree that passing the information as structures or callbacks for TPA and related properties is a good idea. But we'll need to continue to support hosts that provide TPA in the original form. So it may introduce a bit of complexity, which is best handled independently.

Reply:
> The TPA-binder will look for an assembly in TRUSTED_BUNDLED_ASSEMBLIES first and use a special call-back if found.

I do not understand what is the benefit of having both `TRUSTED_BUNDLED_ASSEMBLIES` and the callback. The callback alone should be enough, no?

Reply:
So, the host keeps `TRUSTED_BUNDLED_ASSEMBLIES` on it's side and returns a pointer to the PE image if found, right? Yes, that will work.

Reply:
Also, nit - I would not use `TRUSTED` in any new names. It is left-over from Silverlight and there is nothing trusted about these in .NET Core.

Reply:
Thanks @Jan, I've updated the PR with these notes in this change: https://github.com/dotnet/designs/pull/74/commits/7c2e77a525b071f573ccc5866545d5d7ed6bd519

Reply:
Great job, @swaroop-sridhar